src/adapter/src/coord/cluster_scheduling.rs

```diff
@@ -0,0 +1,169 @@
+
+impl Coordinator {
+    #[mz_ore::instrument(level = "debug")]
+    /// Call each scheduling policy, gather their decisions about turning clusters on/off, and send
+    /// a message with the result.
+    pub(crate) async fn check_scheduling_policies(&mut self) {
+        // (So far, we have only this one policy.)
+        let refresh_policy_decisions = self.check_refresh_policy().await;
+
+        let internal_cmd_tx = self.internal_cmd_tx.clone();
+        if let Err(e) = internal_cmd_tx.send(Message::SchedulingDecisions(vec![(
+            REFRESH_POLICY_NAME,
+            refresh_policy_decisions,
+        )])) {
+            // It is not an error for this task to be running after `internal_cmd_rx` is dropped.
+            warn!("internal_cmd_rx dropped before we could send: {:?}", e);
+        }
+    }
+
+    async fn check_refresh_policy(&mut self) -> Vec<(ClusterId, bool)> {
+        let mut decisions: Vec<(ClusterId, bool)> = Vec::new();
+        for cluster in self.catalog().clusters() {
+            if let ClusterVariant::Managed(ref config) = cluster.config.variant {
+                match config.schedule {
+                    ClusterScheduleOptionValue::Manual => {
+                        // nothing to do, user manages it manually
+                    }
+                    ClusterScheduleOptionValue::Refresh => {
+                        let refresh_mvs = cluster
+                            .bound_objects()
+                            .iter()
+                            .filter_map(|id| {
+                                if let CatalogItem::MaterializedView(mv) =
+                                    self.catalog().get_entry(id).item()
+                                {
+                                    if mv.refresh_schedule.is_some() {
+                                        Some(id)
+                                    } else {
+                                        None
+                                    }
+                                } else {
+                                    None
+                                }
+                            })
+                            .collect_vec();
+
+                        // Check whether there is at least one REFRESH MV needing a refresh.
+                        let local_read_ts = self.get_local_read_ts().await;
+                        decisions.push((
+                            cluster.id,
+                            refresh_mvs.into_iter().any(|mv| {
+                                let mv_frontier = &self
+                                    .controller
+                                    .storage
+                                    .collection(*mv)
+                                    .expect("the storage controller should know about MVs that exist in the catalog")
+                                    .write_frontier;
+                                mv_frontier.less_than(&local_read_ts)
+                            }),
+                        ));
```

Comment:
I'd recommend sprinkling some `debug`/`trace` logs in here so if this logic goes off the rails, we can get some insight into why.

Reply:
Done: 10d8434dee95a3c2dead88ccb579487eb3be0b7a